### ISSUE TITLE
`import torch` results in `cuInit` call

### ISSUE THREAD

**Issue by malfet** [pytorch/pytorch#116276]
### 🐛 Describe the bug

Consider the following:
```
$ lldb -- python3 -c "import torch;"
(lldb) target create "python3"
Current executable set to 'python3' (x86_64).
(lldb) settings set -- target.run-args  "-c" "import torch;"
(lldb) b cuInit
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) r
Process 3597474 launched: '/usr/bin/python3' (x86_64)
Process 3597474 stopped and restarted: thread 1 received signal: SIGCHLD
1 location added to breakpoint 1
warning: (x86_64) /home/nshulga/.local/lib/python3.8/site-packages/numpy.libs/libgfortran-040039e1.so.5.0.0 No LZMA support found for reading .gnu_debugdata section
Process 3597474 stopped
* thread #1, name = 'python3', stop reason = breakpoint 1.1
    frame #0: 0x00007fff8c6b59c0 libcuda.so.1`cuInit
libcuda.so.1`cuInit:
->  0x7fff8c6b59c0 <+0>:  cmpl   $0x321cba00, 0x19dafee(%rip) ; imm = 0x321CBA00 
    0x7fff8c6b59ca <+10>: je     0x7fff8c6b59d8            ; <+24>
    0x7fff8c6b59cc <+12>: jmpq   *0x18e2db6(%rip)
    0x7fff8c6b59d2 <+18>: nopw   (%rax,%rax)
```
With 2.1.2 it does not happen:
```
$ lldb -- python -c "import torch"
(lldb) target create "python"
Current executable set to 'python' (x86_64).
(lldb) settings set -- target.run-args  "-c" "import torch"
(lldb) b cuInit
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/bin/python unsupported DW_FORM value: 0x1f
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) r
Process 3598562 launched: '/home/nshulga/miniconda3/envs/py311/bin/python' (x86_64)
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/math.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_ctypes.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_struct.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_opcode.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_typing.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/zlib.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_bz2.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_lzma.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_bisect.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_random.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_sha512.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/libstdc++.so.6 unsupported DW_FORM value: 0x1f
1 location added to breakpoint 1
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_json.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/site-packages/numpy.libs/libgfortran-040039e1.so.5.0.0 No LZMA support found for reading .gnu_debugdata section
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_datetime.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_pickle.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_contextvars.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/binascii.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_hashlib.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_blake2.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_heapq.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/grp.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/fcntl.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_posixsubprocess.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/select.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/cmath.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_socket.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/array.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_multiprocessing.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_uuid.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_ssl.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/unicodedata.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_queue.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_decimal.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
warning: (x86_64) /home/nshulga/miniconda3/envs/py311/lib/python3.11/lib-dynload/_asyncio.cpython-311-x86_64-linux-gnu.so unsupported DW_FORM value: 0x1f
Process 3598562 exited with status = 0 (0x00000000) 
```
Likely solved by https://github.com/pytorch/pytorch/pull/116201

### Versions

nightly, 2.2.0

cc @ezyang @gchanan @zou3519 @kadeng @ptrblck @seemethere @pytorch/pytorch-dev-infra

**Comment by xuzhao9** [comment]
This could also be related to https://github.com/pytorch/benchmark/issues/2064

**Comment by malfet** [comment]
I guess I should add the following test to nightlies:
```
lldb --batch -o "b cuInit" -o "run" -- python3 -c "import torch;print(torch.cuda.device_count() > 0)"
```

**Comment by ptrblck** [comment]
@malfet could this be related to https://github.com/pytorch/pytorch/pull/112623 which still breaks Nsight Systems profiling?
Can we revert it now or do you plan on more debugging steps?

**Comment by malfet** [comment]
@ptrblck it's already been reverted, but looking for a test to avoid this in the future.

**Comment by ptrblck** [comment]
@malfet Ah, great! Sorry for missing it in the first comment.

**Comment by xuzhao9** [comment]
cc the original commit author @aaronenyeshi for visiability.

**Comment by speediedan** [comment]
Thought it might be worth noting that https://github.com/pytorch/pytorch/pull/112623 was also breaking most DDP fork use cases as well (used by downstream packages like Lightning, [Fine-Tuning Scheduler](https://github.com/speediedan/finetuning-scheduler/) etc.). Thanks for finding and reverting this for 2.2.0 everyone!

I'd love to be kept in the loop regarding the approach to refactoring https://github.com/pytorch/pytorch/pull/112623 if you don't mind (I'll try to keep an eye for the relevant PRs but  @ me if you think of it).

Thank you all for your invaluable contributions to the open-source ML community. I know you may too often hear "bug, fix plz" from the community, but your collective contributions are immensely valuable and you should feel commensurate appreciation. Excited for what's on the horizon! Thanks guys!

 @xuzhao9 @malfet @ptrblck @aaronenyeshi

**Comment by malfet** [comment]
Great idea from @Aidyn-A : use ctypes to dlopen and make a cuDeviceGetCount API call, which should fail if `cuInit` has not been called previously. Will take an AI to implement such unit test

**Comment by atalman** [comment]
validated final rc
```
lldb --batch -o "b cuInit" -o "run" -- python3 -c "import torch;print(torch.cuda.device_count() > 0)"
```
